### ISSUE TITLE
Adding grpc-all:1.35.0 to Kotlin gradle project breaks classpath

### ISSUE THREAD

**Issue by 0xbeefbeef** [grpc/grpc-kotlin#232]
### Adding grpc-all to vanilla Kotlin gradle project seems to break the classpath

Minimal example is as follows:

**build.gradle:**

```
plugins {
    id "org.jetbrains.kotlin.jvm"
    id 'kotlin'
}

apply plugin: 'application'

mainClassName = 'com.example.lib.MyClass'

jar {
    manifest {
        attributes 'Main-Class': mainClassName
    }
    from {
        configurations.runtimeClasspath.collect { it.isDirectory() ? it : zipTree(it) }
    }
}

java {
    sourceCompatibility = JavaVersion.VERSION_1_8
    targetCompatibility = JavaVersion.VERSION_1_8
}

dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
    implementation "io.grpc:grpc-all:1.35.0"
}
```

**MyClass.kt**:
```
package com.example.lib

class MyClass {
    companion object {
        @JvmStatic fun main(args: Array<String>) {
            println("Hello World!")
        }
    }
}
```

The project builds without any errors. However, when starting the resulting jar I get an error:

`Error: Could not find or load main class com.example.lib.MyClass. Reason: java.lang.ClassNotFoundException: 
com.example.lib.MyClass`

Removing the line `implementation "io.grpc:grpc-all:1.35.0"` from the build.gradle solves this. Does anyone have an idea what might cause this behavior?

**Comment by jamesward** [comment]
Not sure. But why do you have the `jar` section? What if you take that out?

**Comment by 0xbeefbeef** [comment]
I think I need at least the manifest part to define the main entry point / main function for the jar. If I leave out the part:
```
  from {
    configurations.runtimeClasspath.collect { it.isDirectory() ? it : zipTree(it) }
  }
```
then I get another error:
```
/data/android-studio-canary/android-studio/jre/bin/java -Dfile.encoding=UTF-8 -jar /home/flo/Code/KotlinTest/lib/build/libs/lib.jar
Exception in thread "main" java.lang.NoClassDefFoundError: kotlin/jvm/internal/Intrinsics
	at com.example.lib.MyClass$Companion.main(MyClass.kt)
	at com.example.lib.MyClass.main(MyClass.kt)
Caused by: java.lang.ClassNotFoundException: kotlin.jvm.internal.Intrinsics
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 2 more
```
I already tried to go without grpc-all, and instead use `grpc-netty-shaded` and `grpc-kotlin-stub` It builds and runs fine, but I have issues with *Android Studio* complaining about unresolved references, and missing or duplicate dependencies in classpath (regarding `NettyChannelBuilder`, or `CallCredentials` for example). I was not able to resolve this, so I tried with grpc-all.

**Comment by jamesward** [comment]
Ah so you are packaging a JAR and trying to run it? The classpath stuff is probably going to be tricky & not portable in that case. Are you open to another way to handle the packaging?

**Comment by 0xbeefbeef** [comment]
For sure. I don't have alot of experience with java/kotlin/jvm. So good ideas are welcome ;)
The application is supposed to run inside a docker container. Using the packaging I did before I got it running, somehow. But I am sure this is not the best solution.

**Comment by jamesward** [comment]
Cool. A couple things... The easy way to containerize is with [Buildpacks](https://buildpacks.io).  Here is a full sample you can start with:
https://github.com/GoogleCloudPlatform/kotlin-samples/tree/master/run/grpc-hello-world-gradle

To containerize that one:

1. [Install pack](https://buildpacks.io/docs/tools/pack/)
2. Build the container:
```
pack build grpc-hello-world-gradle --builder gcr.io/buildpacks/builder
```
3. Run it:
```
docker run -it grpc-hello-world-gradle
```

For something that more matches what you have above, I created a repo you can reference:
https://github.com/jamesward/hello-grpc-all

**Comment by 0xbeefbeef** [comment]
Thank you James! Sounds really interesting for me. Are the resulting image sizes comparable to the Dockerfile approach?

**Comment by jamesward** [comment]
Yeah, the images are pretty similar to manually created ones.

**Comment by 0xbeefbeef** [comment]
While trying to use *buildpacks* to build my docker images, I ran into a few issues with my setup. I think this is mainly caused by the fact that I am an absolute gradle newbie, and my project setup (with several subprojects) was probably not very well configured. There were some dependencies the pack cli tool could not resolve properly and which I was not able to resolve. 

Anyhow, I now gave *jib* a chance. Et voila, it works perfectly out of the box. Very small images, smart layering. Any reason I should nevertheless try to get the *buildpack* solution working?
